Issue #13345: Enable Test for TranslationCheckExamplesTest

Broilzzz opened the pull request:
Issue #13345

im having trouble with this error, i tried looking at other peoples merged pull requests but wasnt able to solve this TreeWalker error.
any help would be helpful

![image](https://github.com/user-attachments/assets/19fdfc89-bd10-47ec-a48a-600853e83743)

romani commented:
Please review this code changes to see direction https://github.com/checkstyle/checkstyle/commit/a79d997a50f6f1413bfc8886bb0063e5a2b646d5

And at https://github.com/checkstyle/checkstyle/pull/16161

Broilzzz commented:
> Please review this code changes to see direction [a79d997](https://github.com/checkstyle/checkstyle/commit/a79d997a50f6f1413bfc8886bb0063e5a2b646d5)
> 
> And at #16161

i have gone through their pull request and done changes accordingly
but the TreeWalker error still persists

![image](https://github.com/user-attachments/assets/82e91b8e-53c7-4ab6-8843-b42817ccf2bc)

romani commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -180,17 +184,19 @@ ButtonLabels_fr.properties: Key 'cancel' is missing.
           Example:
         </p>
         <source>
+# // violation below 'name, cancel and age' is missing
```
There should be macros, to get content from file

romani replied:
My bad, it is final xml, not a template.

romani commented on src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckExamplesTest.java:
```diff
@@ -33,28 +34,45 @@ protected String getPackageLocation() {
     @Test
     public void testExample1() throws Exception {
+        final String configFilePath = getPath("Example1.java");
+        final String propertyFilePath = getPath("Example1.properties");
```
You should have few files, as translation files are usually in different files, not a single file .
Do array of files, to let `verifyWithInlineConfigParserSeparateConfigAndTarget` run on array (few files)

romani reviewed (changes requested):
Items

romani commented:
GitHub, generate website

Broilzzz commented:
Here is the changes i have done locally but not yet push to origin

Example1Messages.properties
`/*xml
<module name="Checker">
  <module name="Translation">
    <property name="fileExtensions" value="properties, translations"/>
    <property name="requiredTranslations" value="fr"/>
  </module>
</module>
*/
// xdoc section -- start
#messages.properties
hello=Hello
cancel=Cancel
// xdoc section -- end
`

Example1Messages.translations
`/*xml
<module name="Checker">
  <module name="Translation">
    <property name="fileExtensions" value="properties, translations"/>
    <property name="requiredTranslations" value="fr"/>
  </module>
</module>
*/
// xdoc section -- start
#messages.translations
hello=Hallo
ok=OK
// xdoc section -- end
`

Example1.java
`package com.puppycrawl.tools.checkstyle.checks.translation;

public class Example1 {
}
`

TranslationCheckExampleTest.java
`package com.puppycrawl.tools.checkstyle.checks;

import org.junit.jupiter.api.Test;
import static com.puppycrawl.tools.checkstyle.checks.TranslationCheck.MSG_KEY_MISSING_TRANSLATION_FILE;
import java.util.List;

import com.puppycrawl.tools.checkstyle.AbstractExamplesModuleTestSupport;


public class TranslationCheckExamplesTest extends AbstractExamplesModuleTestSupport {
    @Override
    protected String getPackageLocation() {
        return "com/puppycrawl/tools/checkstyle/checks/translation";
    }

    @Test
    public void testExample1() throws Exception {
        final String file1 = getPath("Example1Messages.properties");
        final String file2 = getPath("Example1Messages.translations");
        final String[] expected = {
            "1: " + getCheckMessage(MSG_KEY_MISSING_TRANSLATION_FILE, "messages_fr.properties"),
            "1: " + getCheckMessage(MSG_KEY_MISSING_TRANSLATION_FILE, "messages_fr.translations"),
        };
        final List<String> expectedFromFile1 = List.of(
                "1: " + getCheckMessage(MSG_KEY_MISSING_TRANSLATION_FILE, "messages_fr.properties")
        );

        final List<String> expectedFromFile2 = List.of(
            "1: " + getCheckMessage(MSG_KEY_MISSING_TRANSLATION_FILE, "messages_fr.translations")
        );
        verifyWithInlineConfigParser(file1, file2, expected);
    }`

But Still im getting this error
![image](https://github.com/user-attachments/assets/c7e6abcc-bea0-47c5-862e-2901473c5199)

romani commented:
please consider following approach to write test method - https://github.com/checkstyle/checkstyle/pull/16161#discussion_r1921632769

we take config from Java file, just because it is limitation of our test suite, and we take all files in folder as target and we have collettion of violations for each file.

romani commented:
@Broilzzz , https://github.com/checkstyle/checkstyle/pull/16161 is merged, we do your PR by example of that PR.

Broilzzz commented:
> @Broilzzz , https://github.com/checkstyle/checkstyle/pull/16161 is merged, we do your PR by example of that PR.

Ok I'm on it

Broilzzz commented:
I figured it out and the changes have been made, please verify the changes 👍

romani commented:
GitHub, generate website

romani commented on src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckExamplesTest.java:
```diff
@@ -33,28 +40,83 @@ protected String getPackageLocation() {
-        final String[] expected = {
+        final DefaultConfiguration checkConfig = createModuleConfig(TranslationCheck.class);
+        checkConfig.addProperty("requiredTranslations", "fr");
+        checkConfig.addProperty("fileExtensions", "properties, translations");
```
config should be loaded from Example file, please look at https://github.com/checkstyle/checkstyle/blob/6422f4d93d6d77898edeb1e9c56a7c9295ef0fb8/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheckExamplesTest.java#L61 for example.

Broilzzz replied:
> config should be loaded from Example file, please look at
> 
> https://github.com/checkstyle/checkstyle/blob/6422f4d93d6d77898edeb1e9c56a7c9295ef0fb8/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpOnFilenameCheckExamplesTest.java#L61
> 
> for example.

The problem when i do that is the TreeWalker becomes the parent of TranslationCheck

Broilzzz replied:
I have made the changes for Example1, do suggest some possible solutions

romani replied:
Can we create new version of `createChecker` that puts config under Checker ? We are not limited to extend existing suite.

Broilzzz replied:
ok ill look into it

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -93,15 +93,15 @@ messages.properties: Key 'ok' missing.
           <param name="path"
-                 value="resources/com/puppycrawl/tools/checkstyle/checks/translation/Example1.txt"/>
+                 value="resources/com/puppycrawl/tools/checkstyle/checks/translation/Example1.java"/>
           <param name="type" value="config"/>
```
please review update https://github.com/checkstyle/checkstyle/pull/16141/files#diff-1a0e99bbb8140dd6852b3339803fbdbf9b106347f8a55ee9cc93f38f87995b7dR54

to see how to use "raw" type and put in web site content of whole file of properties.

romani reviewed (changes requested):
items:

Broilzzz commented:
All Changes have been made, I noticed that the CHECKER_CHILDREN in TestInputConfiguration did not contain TranslationCheck which resulted in TreeWalker becoming the parent of TranslationCheck.
I have added this line to it which fixed the problem

also added format changes for translation xml file

romani commented:
GitHub, generate website

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -93,17 +93,29 @@ messages.properties: Key 'ok' missing.
         </macro>
         <p id="Example1-code">
-          Example:
+          Example1:
```
It took me a while to understand first example.....

please add one more block here to explain content of folder (attention no messages.translations ):
application.properties
messages.properties

please change property to `<property name="fileExtensions" value="properties"/>` 

this example will show how Check differentiate config files and transaltion files.

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -93,17 +93,29 @@ messages.properties: Key 'ok' missing.
+          Contents of messages.properties file:
+        </p>
+        <source>
+hello=Hello
```
please review update https://github.com/checkstyle/checkstyle/pull/16141/files#diff-1a0e99bbb8140dd6852b3339803fbdbf9b106347f8a55ee9cc93f38f87995b7dR54

to see how to use "raw" type and put in web site content of whole file of properties.
content shoudl not be hardcoded in template, it should be loaded from property file.

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -114,17 +126,29 @@ messages.translations: Properties file 'messages_fr.translations' is missing.
         </macro>
         <p id="Example2-code">
-          Example:
+          Example2:
```
please apply print of folder content to all Examples.
Print of files in folder can be placed to Input file and loaded to template by "raw" type macros

see example at https://github.com/checkstyle/checkstyle/commit/d2da459f3e1de3464160e308b82abaf7b58c8e0c#diff-6cc6ac93b182101bb7cbcffff9f97f80a68b7a8df13483dac2ec65db7cc3b3fbR9

in this case Input file has config and visual representation of validation, it is easier to maintain.

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/translation/Example3.java:
```diff
@@ -6,6 +6,8 @@
 </module>
 */
 
```
please swap this Example3 with Example1.
Ideally we should start introduction of user to Check with minimal set of properties, to show how it work only by one customization from default.

Broilzzz replied:
> please swap this Example3 with Example1. Ideally we should start introduction of user to Check with minimal set of properties, to show how it work only by one customization from default.

ok ill swap example3 to example1 and rename and reformat both of them

romani reviewed (changes requested):
good, items to improve:

Broilzzz commented:
changes done as requested
please verify

romani commented:
GitHub, generate website

romani commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -115,32 +115,59 @@ messages.properties: Key 'ok' missing.
+        </p>
+        <source>
+--&gt; Example1
+  |--&gt; messages.properties
```
Please make whole source block like:
```
messages.properties
messages_ja.properties
messages_fr.properties
```
Flat is good, user should not kare what is parent folder.

romani reviewed (changes requested):
Please do for all:

romani commented on src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckExamplesTest.java:
```diff
@@ -33,28 +42,89 @@ protected String getPackageLocation() {
     public void testExample2() throws Exception {
-        final String[] expected = {
-
+        final Map<String, List<String>> expected = new HashMap<>();
```
Move this line close to first usage of this variable.

Broilzzz replied:
done

romani commented on src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckExamplesTest.java:
```diff
@@ -33,28 +42,89 @@ protected String getPackageLocation() {
-
+        final Map<String, List<String>> expected = new HashMap<>();
+        final File[] propertyFiles = {
+            new File(getPath("ButtonLabels.properties")),
```
`getPath("ButtonLabels.properties")` this happens twice in this method.
Please optimize code to do it ones.

Broilzzz replied:
done

romani reviewed (changes requested):
Items

Broilzzz commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -115,32 +115,59 @@ messages.properties: Key 'ok' missing.
-messages.properties: Properties file 'messages_fr.properties' is missing.
-messages.translations: Properties file 'messages_fr.translations' is missing.
+greeting=?????
+age=??
```
changed japanese  to spanish

romani commented:
GitHub, generate website

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/translation/Example1.java:
```diff
@@ -0,0 +1,18 @@
+// xdoc section -- start
+/*
+--> Example3
+  |--> messages.properties
```
Please put `// xdoc section -- start` after /*
And load this content as "code" in template.
Flattening is required to be here.

Broilzzz replied:
but if the code content is brought, can we make the (file-structure) inside the Example class itself and not make a new block just for (file-content) ?

Broilzzz replied:
> Please put `// xdoc section -- start` after /* And load this content as "code" in template. Flattening is required to be here.

but if the code content is brought, can we make the (target folder structure) inside the Example class itself and not make a new block just for (target folder structure) ?

romani replied:
content for website file should be already in xml file.
We do not have validation of violations on files in our test suite for now, I hope one day we will do this, but for now lets just enable Example with a little of copy paste and manual validation of content.
and in next issue/PR we can make such blocks to be validated against actual validation results.

romani reviewed (changes requested):
Items

romani commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -115,32 +115,58 @@ messages.properties: Key 'ok' missing.
-          Example:
+        <p id="Example1-raw">
+          Example1:
+          (file-structure):
```
(target folder content)

Broilzzz replied:
done

romani replied:
@Broilzzz , not done, please do. Do for all such text in your PR

romani reviewed (changes requested):
Items

romani commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -154,55 +180,76 @@ messages.translations: Properties file 'messages_fr.translations' is missing.
+          (file-structure):
+        </p>
+        <source>
+messages.properties (attention, no messages_fr.properties file)
```
Move `(attention` to P tag above, to be part of explanation of source block

Broilzzz replied:
done

romani reviewed (changes requested):
Items

romani commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -154,55 +180,76 @@ messages.translations: Properties file 'messages_fr.translations' is missing.
+ok=OK
+        </source>
+        <p id="Example3-validation">
+          Validation for Example1:
```
`Validation for Example3:`

Broilzzz replied:
done

romani reviewed (changes requested):
Items

romani commented:
Please apply this patche to your commit 
https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=25883&view=logs&j=6a833393-c551-5000-0916-5000fd3b86e7&t=bdbbd129-77f0-5d73-ce7d-db58d34c8358&l=48

romani commented:
GitHub, generate website

romani commented:
GitHub, generate website

Broilzzz commented:
> Please apply this patche to your commit https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=25883&view=logs&j=6a833393-c551-5000-0916-5000fd3b86e7&t=bdbbd129-77f0-5d73-ce7d-db58d34c8358&l=48
> 
> Please fix this inspection violation https://app.circleci.com/pipelines/github/checkstyle/checkstyle/30041/workflows/94c6e2c9-0a2f-45ae-8e9a-579c1f58cb62/jobs/778366/artifacts

applied the patch

romani commented:
GitHub, generate website

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/translation/Example1.java:
```diff
@@ -0,0 +1,20 @@
+
+package com.puppycrawl.tools.checkstyle.checks.translation;
+
+// xdoc section -- start
+public class Example1 {}
+/*
+(target folder structure):
```
`// xdoc section -- start` should under `/*`

content between `// xdoc section -- xxxxx` is copied to final xml file in xdoc folder that will be source for html.

We do here small cheating we put non java code to java file to reuse `type="code"`

romani replied:
after update please run `mvn clean site -Pno-validations` on your local https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-generate-website-only

to validate that result html is good.

Broilzzz replied:
is this fine?
```
package com.puppycrawl.tools.checkstyle.checks.translation;

/*
// xdoc section -- start
(target folder structure):

messages.properties
messages_fr.properties
messages_es.properties
*/
public class Example1 {}
// xdoc section -- end
```

romani replied:
```

/*
// xdoc section -- start
(target folder structure):

messages.properties
messages_fr.properties
messages_es.properties
// xdoc section -- end
*/
public class Example1 {}
```

Broilzzz replied:
ill add the (target folder structure) inside the p tag in the xml file

romani reviewed (changes requested):
items:

romani commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -155,54 +181,78 @@ messages.translations: Properties file 'messages_fr.translations' is missing.
+        </p>
+        <source>
+(attention, no messages_fr.properties file)
+(attention, no messages_fr.translations file)
```
Make it `Example3 no messages_fr.* files (target folder structure):`

Broilzzz replied:
done

romani commented on src/site/xdoc/checks/misc/translation.xml:
```diff
@@ -155,54 +181,78 @@ messages.translations: Properties file 'messages_fr.translations' is missing.
+(attention, no messages_fr.translations file)
+
+messages.properties
+messages.translations
```
Put as trailing comment violations that Check populated

romani reviewed (changes requested):
Items

romani commented:
GitHub, generate website

romani commented:
Please fix inspection violation https://app.circleci.com/pipelines/github/checkstyle/checkstyle/30064/workflows/51ad4b9f-403e-4d73-b958-010bf48962d3/jobs/779332/artifacts

```
<problem>
<file>file://$PROJECT_DIR$/src/site/xdoc/checks/misc/translation.xml</file>
<line>136</line>
<module>project</module>
<entry_point TYPE="file" FQNAME="file://$PROJECT_DIR$/src/site/xdoc/checks/misc/translation.xml"/>
<problem_class id="XmlDuplicatedId" severity="ERROR" attribute_key="ERRORS_ATTRIBUTES">Duplicate 'id' attribute</problem_class>
<description>Duplicate id reference</description>
<highlighted_element>messages-raw</highlighted_element>
<language>XML</language>
<offset>15</offset>
<length>12</length>
</problem>
<problem>
<file>file://$PROJECT_DIR$/src/site/xdoc/checks/misc/translation.xml</file>
<line>236</line>
<module>project</module>
<entry_point TYPE="file" FQNAME="file://$PROJECT_DIR$/src/site/xdoc/checks/misc/translation.xml"/>
<problem_class id="XmlDuplicatedId" severity="ERROR" attribute_key="ERRORS_ATTRIBUTES">Duplicate 'id' attribute</problem_class>
<description>Duplicate id reference</description>
<highlighted_element>messages-raw</highlighted_element>
<language>XML</language>
<offset>15</offset>
<length>12</length>
</problem>
<problem>
<file>file://$PROJECT_DIR$/src/site/xdoc/checks/misc/translation.xml</file>
<line>243</line>
<module>project</module>
<entry_point TYPE="file" FQNAME="file://$PROJECT_DIR$/src/site/xdoc/checks/misc/translation.xml"/>
<problem_class id="XmlDuplicatedId" severity="ERROR" attribute_key="ERRORS_ATTRIBUTES">Duplicate 'id' attribute</problem_class>
<description>Duplicate id reference</description>
<highlighted_element>messages-raw</highlighted_element>
<language>XML</language>
<offset>15</offset>
<length>12</length>
</problem>
</problems>
```

romani commented:
Ok, please fix all above and we will be good to merge

Broilzzz commented:
> Ok, please fix all above and we will be good to merge

done, please verify

romani commented:
GitHub, generate website

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -88,70 +88,141 @@ messages.properties: Key 'ok' missing.
-          Example:
+          Example1
+          (target folder structure):
         </p>
```
please make P tag single line, you have panty of space.

do same for all others.

Broilzzz replied:
done

romani replied:
not done, please do
`<p id="Example1-file"> Example1 (target folder structure):</p>`

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -88,70 +88,141 @@ messages.properties: Key 'ok' missing.
-messages.properties: Properties file 'messages_fr.properties' is missing.
-messages.translations: Properties file 'messages_fr.translations' is missing.
+messages.properties: Key 'age' missing.
+messages.properties: Key 'name' missing.
```
please load such content from ExampleX.java files. `<param name="type" value="code"/>` 

do this for all.

Broilzzz replied:
since im already using the code param macro to get the (target folder structure), should i make a new java file? or hard code the (target file structure)?

romani replied:
Same java file. 
src/site/xdoc/checks/misc/translation.xml.template should have almost no hardcoded text for examples, all should be loaded from files that are active in test execution.

romani replied:
In other words, this lines should be removed and on their place there should be macros that loads such content from java file.

Broilzzz replied:
if we take it from the same java file, then we are not able to make two different blocks
ie, (target file structure) block that displays the file structure and a validation block that displays the error
we could hard code the file structure and use macro for the error

we can use macro to get the validation like this
```
messages.properties // violation, 'Key 'greeting', 'age' and 'name' missing.'
messages_fr.properties // violation, 'Key 'hello', 'cancel' and 'age' missing.'
messages_es.properties // violation, 'Key 'hello', 'cancel' and 'name' missing.'
```

and we can hard code the file structure like this
```
        <p id="Example1">
          Example1 (target folder structure):
        </p>
        <source>
messages.properties
messages_fr.properties
messages_es.properties
        </source>
```


if not is there another type we could use to get the contents from the same java file other than code and config

romani replied:
Ok, I see a problem now.
Let's put list of files as hardcoded and list with validation from java file.

Broilzzz replied:
done

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -88,70 +88,141 @@ messages.properties: Key 'ok' missing.
           extensions will be considered as files that belong to different resource bundles.
```
please remove from Check description example https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a096999_20250209143329/checks/misc/translation.html#Description

we have it in Examples, so it is not required any more.

Broilzzz replied:
done

romani commented on src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckExamplesTest.java:
```diff
@@ -33,28 +42,86 @@ protected String getPackageLocation() {
     public void testExample1() throws Exception {
-        final String[] expected = {
-
+        final Map<String, List<String>> expected = new HashMap<>();
```
move this line to be right above `expected.put(messagesFr,.....`

do same for all others, like this. we should minimized distance from declaration to first usage.

Broilzzz replied:
done

romani commented on src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckExamplesTest.java:
```diff
@@ -33,28 +42,86 @@ protected String getPackageLocation() {
+                InlineConfigParser.parse(configFile);
+        final DefaultConfiguration parsedConfig =
+                testInputConfiguration.createConfiguration();
+        verify(
```
make it like:
```
        final File[] propertyFiles = {
                new File(messages),
                new File(messagesFr),
                new File(messagesEs),
        };

        verify(createChecker(parsedConfig), propertyFiles, expected);
```

distance is reduced, method takes whole horizontal space.


please do this for others

Broilzzz replied:
done

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/translation/Example3.java:
```diff
@@ -0,0 +1,22 @@
+/*
+// xdoc section -- start
+(attention, no messages_home_fr.properties file)
+(attention, no messages_home_fr.translations file)
```
remove this attnetions, they are now part of xml template.

Broilzzz replied:
done

romani commented on src/site/xdoc/checks/misc/translation.xml.template:
```diff
@@ -88,70 +88,141 @@ messages.properties: Key 'ok' missing.
-messages_ja.properties: Key 'cancel' missing.
-messages_ja.properties: Key 'name' missing.
+messages_home.properties: Properties file 'messages_home_fr.properties' is missing.
+messages_home.translations: Properties file 'messages_home_fr.translations' is missing.
```
```
messages_home.properties     // violation 'Properties file 'messages_home_fr.properties' is missing.'
messages_home.translations  // violation 'Properties file 'messages_home_fr.translations' is missing.'
```

lets match violation markers as it is done in other checks, even for this Check it is not required.
please do for all.

Broilzzz replied:
done

romani reviewed (changes requested):
good job !!

last:

romani commented:
@Broilzzz , please resolve conflict

Broilzzz commented:
@romani, all fixes done
please note that in example3 there was word limit exceeded for line so i had to extend it to a new line
```
/*
// xdoc section -- start
messages_home.properties   // violation,
                          'Properties file 'messages_home_fr.properties' missing.'
messages_home.translations // violation,
                          'Properties file 'messages_home_fr.translations' missing.'
// xdoc section -- end
*/
```

romani commented:
GitHub, generate website

romani commented:
GitHub, generate website

romani reviewed (approved):
Awesome. Huge improvement for all users.
